[Ng Tze Wei Bruce Mitchell] iP

ntwbruce opened the pull request:
# Welcome to Duke, your very own task manager!

**Things you can do with ~~Duck~~ Duke:**
1. Add different kinds of tasks
   * _Duke can keep track of tasks with specific deadlines._
   * _Having an event? Duke will keep track of your event periods as well._
   * _Just need a simple to-do list? Duke will do that too._
2. Show that tasks are done
   * _Once a task is done, mark it as so! Duke will keep track of the tasks you have completed._
3. Delete tasks
   * _Done with a task? Get rid of it. Typo? Just delete it and do it again._
4. Find a specific task
   * _Need to know what to get done today? Find tasks by date!_
   * _You can also find tasks by keywords!_

**Why you need Duke:**
- Why worry about forgetting tasks and deadlines?
- Why worry about whether you've finished something?
- I needed to make a bullet list.

All you need to do is: 

- [ ] Head over to this application's [GitHub page](https://github.com/ntwbruce/ip)!
- [ ] Download the JAR file of "Duke v0.1" under Releases
- [ ] Run the JAR file, and you're good to go! :thumbsup:

If you're curious to play around and customise this app for yourself, 
feel free to download the source code from the same [GitHub page](https://github.com/ntwbruce/ip)!

You will find the main function in `src/main/java/duke/Duke`, as shown below.
````java
public class Duke {
    // ...
   public static void main(String[] args) {
      new Duke("data", "duke.txt").run(System.in);
   }
}
````

> See Mumei-ter!

:lemon:

Commit:
Add Gradle support

ntwbruce committed:
Added greet, exit and echo functionality to Duke.java

ntwbruce committed:
Added add and list functionality to Duke.java

ntwbruce committed:
Created new class Task.java

ntwbruce committed:
Modified Duke.java to achieve 'mark as done' functionality with Task objects

ntwbruce committed:
Task list number changed to be by taskList array index rather than by assigned index at Task creation

ntwbruce committed:
Fixed output formatting for Task.java

ntwbruce committed:
Removed redundant Task index attribute

ntwbruce committed:
Moved setDone() print line and taskList periods from Task.java to Duke.java

ntwbruce committed:
Created Todo.java

ntwbruce committed:
Created Deadline.java

ntwbruce committed:
Created Event.java

ntwbruce committed:
Modified Duke.java to handle different types of Tasks

ntwbruce committed:
Fix intro text and close Duke.java Scanner for testing purposes

ntwbruce committed:
Created AutoUITesting test case in input.txt

ntwbruce committed:
Created expected test case outcome in EXPECTED.TXT

ntwbruce committed:
Created DukeException.java and added various potential exception outputs

ntwbruce committed:
Added try-catch block and invalid input exception handling functionality to Duke.java

ntwbruce committed:
Modified Duke and DukeException so that Duke's catch block prints the error message

ntwbruce committed:
Added incomplete task input exceptions to Duke

ntwbruce committed:
Added exceptions for missing deadline time and event period

ntwbruce committed:
Added new exception case and new catch block for NumberFormatException to handle invalid 'done' commands

ntwbruce committed:
Reformatted Duke for greater clarity

ntwbruce committed:
Reformatted Duke for greater clarity again

ntwbruce committed:
Modified Duke to use ArrayList for taskList instead of Task array

ntwbruce committed:
Added invalidDelete case for DukeException

ntwbruce committed:
Added delete functionality with exception handling to Duke

ntwbruce committed:
Modified done and delete functions to throw DukeException when given negative index values

ntwbruce committed:
Created DukeExceptionType.java for enumeration of DukeException types

ntwbruce committed:
Modified DukeException.java to utilise DukeExceptionType enums

ntwbruce committed:
Modified Duke.java to utilise DukeExceptionType enums

ntwbruce committed:
Updated Duke functions to handle unnecessary whitespaces

ntwbruce committed:
Format to follow Java coding standard

There are a few standards that the code fails to follow.

Let's,
* fix the switch case indentation
* change boolean name bye to isBye in Duke class
* remove unnecessary this keywords from Event, Deadline, Task classes

Improves readability and follows the standard.

ntwbruce committed:
Create hard disk file to save task list

Adding the ability to save tasks between separate uses of the program
improves usability, rather than have to start a new task list from
scratch every time.

We start by creating a file 'duke.txt' which acts as the hard disk
in a new directory 'data', and allow Duke to check if the hard disk
exists, create it if not, and access it.

Let's
* create the folder 'data' in the program root folder and the file
'duke.txt' inside 'data' to act as a hard disk and store the task list
* add functionality to Duke to check if 'data' and 'duke.txt' already
exist and create them if they don't already exist
* add exception handling in case the program is unable to create 'data'
and 'duke.txt'

ntwbruce committed:
Create methods for interacting with hard disk

Duke needs methods to interact with the hard disk whenever a change
is made to the task list, allowing the hard disk to keep track of them.

Let's create the methods addEntry, setEntryDone, and deleteEntry to
add, delete and set tasks to done on the hard disk.

After this these methods will be implemented into Duke's main function.

There is also a need to abstract out the process of checking duke.txt
upon starting the program, and adding the functionality of loading in
the tasks saved to the hard disk.

ntwbruce committed:
Implement hard disk interaction methods in Duke

Implement the methods we added in the previous commit in the Duke
function so any changes to the task list in Duke are reflected in the
hard disk as well.

Also, create a custom string format for the tasks added to the hard
disk for readability.

Finally, add exception cases to DukeException to deal with new potential
exceptions.

Let's
* implement addEntry, setEntryDone and deleteEntry in the add, done
and delete logic of the Duke.main() function
* create a new string format for the different tasks for adding to the
hard disk
* implement new exception cases to deal with potential exceptions
arising from the new methods

ntwbruce committed:
Create launchDuke function in Duke class

To improve readability, abstract out the checking of duke.txt at the
start of the Duke.main() function into a function called launchDuke, and
add to it the loading in of the saved tasks from previous sessions.

Let's
* create the launchDuke function to handle the checking of duke.txt
and inputing of the hard disk's saved tasks into Duke's task list
* add exception case in the event the parsing of tasks from the hard
disk runs into an error (if a task was added in a wrong format etc)

ntwbruce committed:
Change Deadline to take in LocalDate parameter

Allow for Deadline to present deadline in a customised format.

Let's
* modify Deadline and Duke to take in a LocalDate parameter for the
deadline
* add an exception case for what the user-provided deadline parameter
is not in an acceptable format

ntwbruce committed:
Allow for time to be included in deadline

Aside from dates, users may want to include a time in their deadline.
This can be achieved by overloading the Deadline constructor to allow
for a LocalTime parameter to be included.

Let's
* modify Deadline and the implementation of adding deadline tasks
in Duke to allow for users to include a deadline time as well
* modify exception case INVALIDDATE to catch DateTimeParseExceptions
arising from the Deadline constructor taking in a LocalTime parameter

ntwbruce committed:
Modify Event class to take in LocalDate and LocalTime parameters

Give users the option of providing dates and times for Event tasks in
the form of LocalDate and LocalTime parameters.

Let's
* overload the Event constructor to allow for the following formats
of Event periods:
1) Start and end dates
2) Start and end times on one day
3) Start date and time, end date and time
* add exception case to deal with the potential exception of invalid
date/time formats

ntwbruce committed:
Add method isTodayTask for date-based task finding

Improve usability by allowing users to find which tasks have a deadline
or event on a certain specified day. First we create methods in the
various task classes that checks if a deadline or event takes place on
that day.

Let's
* convert Task to an abstract class with the abstract isTodayTask method
* implement isTodayTask in Deadline, Event and Todo

ntwbruce committed:
Implement date-based find function in Duke

Carrying on from previous commit, creating a find function allows users
to find deadlines or events that take place on a specified day.

Let's
* implement the find functionality in Duke.main()
* create a new exception case to handle the case when users don't give
a date or given a date in the wrong format in the find function

ntwbruce committed:
Merge branch 'branch-Level-7'

ntwbruce committed:
Merge branch 'branch-Level-8'

ntwbruce committed:
Create new classes to extract from Duke

In line with the Object-Oriented Programming paradigm, extract these
various classes from Duke so they can be worked on individually and
Duke doesn't become bloated.

Let's create new classes Storage, TaskList and Ui in preparation of
refactoring Duke to reduce its size and improve readability.

ntwbruce committed:
Implement the Ui class

Extract all the parts of Duke that involve output to the console and
delegate them to the Ui class.

Let's
* implement methods in the Ui class to replace the console output
commands in Duke, so whenever there is a need to print something to
the console, it is handled by the Ui class
* assign the Duke class a Ui attribute with which the message printing
methods are called

ntwbruce committed:
Implement the Storage class

Create a Storage class to manage the saving and retrieval of data to and
from the hard disk.

Let's extract the launchDuke, addDBEntry, deleteDBEntry and
setDBEntryDone methods to the Storage class and give Duke a Storage
attribute to allow it to manage storage with the hard disk.

ntwbruce committed:
Implement the TaskList class

Implement TaskList to handle any and all functionality to do with
altering the task list. In addition, TaskList will also handle the
changes made to the storage of tasks by Storage.

Let's extract any functionality to do with the task list (methods to
add, delete, make changes to the task list, as well as UI console output
methods upon task list changes) for the TaskList class to handle.

ntwbruce committed:
Implement the Parser class

The Parser class takes over the Duke class's job of taking in an
instruction from the user, interpreting it and performing the relevant
action after.

It will either return a Command (to be implemented) for the Duke class
to perform, or throw an exception to the Duke class to handle.

Let's implement the Parser class to parse user input and return a
Command to be executed.

ntwbruce committed:
Implement the Command class

A Command object takes in parsed user input from Parser and performs the
relevant task (exiting the program, showing list, task list functions,
etc).

Let's create and implement a Command class that Duke will use to
execute commands from the user.

ntwbruce committed:
Refactor Duke to utilise Command and Parser

Now that Command and Parser have been implemented, we can refactor Duke
to utilise them.

Let's edit out the Duke logic that has been passed to Command and Parser
and utilise those classes in the implementation of Duke's run function.

ntwbruce committed:
Extract different commands from the Command class

For greater clarity, the Command class can be made into an abstract
class, and create more specific Command classes (eg. AddCommand) to
handle each type of command.

Let's
* create new classes: AddCommand, ByeCommand, DeleteCommand,
DoneCommand, FindCommand and ListCommand, that extend Command, which is
made abstract, and extract the functionality from the original class
to the child classes
* modify Parser to use the new classes

ntwbruce committed:
Organise into Packages

Create packages within the project's java folder to organise the various
classes for easier management.

Let's
* Create several packages to sort our classes into
* Refactor and ensure correct import statements for all classes

ntwbruce committed:
Make minor formatting changes

At this juncture make minor modifications to improve the look of the
code and make amendments to adhere to coding standards.

Let's
* fix the names of DukeExceptions to follow coding standard
* extend Ui console output lines

ntwbruce committed:
Add JUnit to the project

In order to partially automate testing of the program, introduce JUnit
to the project to allow for organised testing of the program's various
parts.

Let's
* create a test subfolder within the project's src folder with which
to add JUnit support to the project
* Add a file DukeTest with a few test cases to test various
functionalities of the main Duke program

ntwbruce committed:
Create StorageTest to JUnit test the Storage class

Partially automatic the testing of the Storage class and its functions.

Let's implement a StorageTest class under the test folder, used to test
that the Storage class functions work as intended.

ntwbruce committed:
Add Manifest from JAR file creation

Produce a JAR file for easy distribution of the Duke program. Commit the
manifest file created as a result.

Let's produce a JAR file for the program and save the manifest file.

ntwbruce committed:
Standardising name of program

Change all instances of the original intended name of the program, Duck,
to Duke, which I started using after a while for some reason.

Let's change all 'Duck' to 'Duke'.

If I have the time to, I might change the program name back to Duck at
some point, but that will take time.

ntwbruce committed:
Add JavaDoc for classes in the duke package

Add JavaDoc comments to explain functionality of public methods to aid
users' understanding.

Let's add JavaDoc comments for public methods in the duke package
classes.

ntwbruce committed:
Add JavaDoc for classes in the command package

Add JavaDoc comments to explain functionality of public methods to aid
users' understanding.

Let's add JavaDoc comments for public methods in the command package
classes.

ntwbruce committed:
Add JavaDoc for classes in the exception package

Add JavaDoc comments to explain functionality of public methods to aid
users' understanding.

Let's add JavaDoc comments for public methods in the exception package
classes.

ntwbruce committed:
Add JavaDoc for classes in the task package

Add JavaDoc comments to explain functionality of public methods to aid
users' understanding.

Let's add JavaDoc comments for public methods in the task package
classes.

ntwbruce committed:
Improve the naming convention of DukeExceptions

Rewrite some exception names for greater clarity.

Let's modify some DukeException names to be clearer, rearrange them for
greater readability.

ntwbruce committed:
Add logical block comments and fix DukeException naming

For greater readability of large logic blocks, split into smaller blocks
and add comments explaining the functionality of each block.

Fix the naming of the DukeException names changed in the previous
commit.

Let's
* include logical block comments
* fix DukeException names

ntwbruce committed:
Create new Command class FindKeywordCommand

In order to implement the functionality off finding tasks by keyword,
create a new command class for it. Differentiate this from the find
command that finds tasks by date.

Let's
* create a FindKeywordCommand class
* rename existing FindCommand class to FindDateCommand
* replace instances of FindCommand to FindDateCommand

ntwbruce committed:
Implement FindKeywordCommand logic in TaskList, Ui and Task classes

The logic of finding tasks by keyword in TaskList calls the Ui class,
which calls the Task class. Hence, modify all three to handle this
logic.

Let's modify the above classes with the functionality of finding tasks
in a task list with a given keyword.

ntwbruce committed:
Update Parser with keyword finding functionality

Parser only handles date finding functionality, so update it to handle
finding tasks with given keyword as well.

Let's add keyword finding functionality to Parser.

ntwbruce committed:
Update Parser and DukeException to handle new exceptions

Handle new exception cases from the newly implemented find tasks by
keyword functionality.

Let's modify Parser, DukeException and DukeExceptionTypes to handle
new exceptions that the find tasks by keyword functionality could
potentially throw.

ntwbruce committed:
Merge branch 'branch-A-JavaDoc'

ntwbruce committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Ui.java

ntwbruce committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/Ui.java
#	src/main/java/exception/DukeException.java
#	src/main/java/exception/DukeExceptionType.java
#	src/main/java/task/Task.java

ntwbruce committed:
Do minor refactoring of duke package classes

There is some unnecessary passing of Ui objects into Storage and
TaskList constructors in Duke when the Ui object can be created in these
classes.

Let's remove unnecessary Ui parameters and allow Storage and TaskList
objects to have their own Ui attributes.

ntwbruce committed:
Remove unnecessary Ui parameter from Command execute method

It's unnecessary. The one child class that does require it, ByeCommand,
can create its own Ui attribute.

Let's remove the Ui parameter from the Command execute method and from
all its child classes.

ntwbruce committed:
Add JavaDoc comments to methods in Test classes

Include JavaDoc comments to aid users' understanding.

Let's explain the functionality of the tests in DukeTest and
StorageTest.

ntwbruce committed:
Add class header comments to all classes

To aid users' understanding of the purpose of each class in the Duke
program.

Let's add JavaDoc comments to every class (because somehow I remembered
to include all the method comments but not the class ones).

ntwbruce committed:
Utilise GFMD in the PR description

For better user understanding, add a nicely stylised information section
in README.md with instructions for setup.

Let's modify README.md to utilise GFMD syntax for better user
readability.

ntwbruce committed:
Add header to README.md

Header not displaying properly on GitHub page, so a syntax fix required.

Let's fix a minor issue with the header syntax in README.md.

ntwbruce committed:
Merge commit 'd8398594b7bc43da5eb865321c5a50cec4b3923d'

ntwbruce committed:
Set up Gradle files

Following the merge of the Gradle support branch, update and add files
to configure Gradle to run and test Duke.

Let's
* set the correct mainClassName in build.gradle
* add checkstyle.xml and suppressions.xml from SE-EDU AddressBook Level
3 project to checkstyle folder in config folder

ntwbruce committed:
Make corrections to existing Test classes

Update test classes to follow current UI output.

Following this, Gradle will be able to correctly run existing unit tests
with the 'test' task provided.

Let's modify DukeTest.java and StorageTest.java to match the earlier
changes made to the Ui class.

ntwbruce committed:
Add enableAssertions to the build.gradle file

This will allows the user of assertions in JUnit tests. Though it does
not affect the current set of tests, it will be useful in the future
when more tests are added that may require it.

Let's enableAssertions in the build.gradle file.

ntwbruce committed:
Merge branch 'branch-gradle'

ntwbruce committed:
Modify suppressions.xml to suppress CustomImportOrder warnings

The standard used by the CustomImportOrder module is different from
CS2103T's coding standard, so we suppress the warnings that are thrown
even though the code follows the standard.

Let's add the CustomImportOrder check to suppressions.xml

ntwbruce committed:
Modify command package classes to follow Checkstyle

Improve code readability by following Java coding standard enforced by
implemented Checkstyle configuration.

Let's fix the formatting errors discovered by running Checkstyle.

ntwbruce committed:
Modify duke package classes to follow Checkstyle

Improve code readability by following Java coding standard enforced by
implemented Checkstyle configuration.

Let's fix the formatting errors discovered by running Checkstyle.

ntwbruce committed:
Modify exception package classes to follow Checkstyle

Improve code readability by following Java coding standard enforced by
implemented Checkstyle configuration.

Let's fix the formatting errors discovered by running Checkstyle.

ntwbruce committed:
Modify task package classes to follow Checkstyle

Improve code readability by following Java coding standard enforced by
implemented Checkstyle configuration.

Let's fix the formatting errors discovered by running Checkstyle.

ntwbruce committed:
Merge branch 'branch-checkstyle'

ntwbruce committed:
Update README.md again

Add more details to the PR description to better convince users to use
Duke (actually the iP tracker didn't update the GFMD tag and I think
it's because bullet list is nested).

Let's create a bullet list that is not nested in a numbered list.

ntwbruce committed:
Set up basic "Hello World" GUI application with JavaFX

Modify the current program to support a GUI with JavaFX for better user
interaction. Start with setting up a simple GUI to test basic
functionality.

Let's
* add JavaFX dependencies to build.gradle
* create Launcher.java
* modify Duke.java to be used by Launcher.java

ntwbruce committed:
Update Duke.java to create GUI window

For Duke to act as an interactive application, it needs a window that
the users interacts with.

Let's add JavaFX logic to create an AnchorPane to produce a window for
Duke.

driip-ddrop commented on README.md:
````diff
@@ -1,24 +1,42 @@
-   | |_| | |_| |   <  __/
-   |____/ \__,_|_|\_\___|
-   ```
+# Welcome to Duke, your very own task manager!
````
hii, I think this should be in your pull request description :)

driip-ddrop commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,185 @@
+                return new AddCommand(newTask);
+            }
+        }
+    }
```
Perhaps implement some other helper methods so that this method is shorter. I noticed this in several other methods as well.

driip-ddrop commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,73 @@
+     *
+     * @return A string representation of this Task's list entry.
+     */
+    public String listEntry() {
```
Maybe a more intuitive name would be better? Eg. displayTask or use toString method

driip-ddrop commented on src/test/java/duke/StorageTest.java:
```diff
@@ -0,0 +1,53 @@
+
+import task.Task;
+
+import java.io.*;
```
Imported classes should be listed explicitly

driip-ddrop commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,95 @@
+ */
+public class TaskList {
+    private final Storage storage;
+    private final ArrayList<Task> list;
```
Instead of "list" as the name, perhaps use names like "tasks" or "taskList" to represent what are the contents of the list ?

driip-ddrop commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,224 @@
+        try {
+            // Create updated file
+            String updatedPath = filePath.split("/")[0] + "/updated.txt";
+            File updated = new File(updatedPath);
```
Maybe use a more descriptive name like updatedFile as it might get confusing in the later lines as to what "updated" is.

driip-ddrop reviewed (commented):
Overall, I think that the code is good :)) There are a few minor issues like naming and methods that are too long but generally the code looks great. good job!!

ntwbruce committed:
Modify Command, TaskList and Ui classes to return Duke response Strings

Originally, with the text-based UI, Duke would System.out.print all of
Duke's responses to user input. Adjust to working with the JavaFX GUI by
now returning the output strings for the GUI to output.

Let's modify these classes so all Commands' execute functions will
return a String representing Duke's response to the user input.

ntwbruce committed:
Create DialogBox.java and amend Duke to handle the new DialogBox class

DialogBox class extracts out the details of creating GUI dialog boxes
which display user interactions with Duke. The handleUserInput and
getResponse methods in Duke must be created to allow the GUI to perform
the Duke logic.

Let's
* create DialogBox.java and add the details for producing dialog boxes
* edit Duke logic to utilise the new DialogBox class, allow the GUI to
run the Duke program.

ntwbruce committed:
Add minor formatting changes

Displaying of messages and lines are formatted wrongly now that the GUI
is being used to display the inputs and outputs rather than the console.

Let's make small edits to these classes to conform with the current
state of the Duke program and the GUI.

ntwbruce committed:
Use FXML to visually enhance GUI

Process of manually enhancing the GUI visuals is tedious, so use FXML
to ease the process.

Let's
* introduce FXML files DialogBox.fxml and MainWindow.fxml
* create MainWindow.java to handle the GUI window
* make changes to Duke, DialogBox and Launcher to accommodate the use
of FXML

ntwbruce committed:
Delete unnecessary methods from Duke

Remove unused methods start (extracted to Main) and showWelcome because
they are unused.

Let's delete the unnecessary methods mentioned above.

ntwbruce committed:
Merge branch 'branch-Level-10'

LiuZhi1an commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,224 @@
+        }
+
+        return savedTasks;
+    }
```
Maybe can consider adding helper methods to make this method shorter instead of having too many nested if-else blocks?

LiuZhi1an replied:
For the load() method😅(sorry for the incomplete selection of lines)

LiuZhi1an commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,73 @@
+ */
+public abstract class Task {
+    private final String taskName;
+    private boolean done;
```
Perhaps can use isDone as boolean name since it is more intuitive as description of the current state? (Though done sounds ok also).

LiuZhi1an commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,189 @@
+        return list.toString();
+    }
+
+    /**
+     * Displays a message to show that the given task has been set to done.
+     *
+     * @param toSetDone The task that has been set to done.
+     * @return String representing the task being done.
+     */
+    public String showDone(Task toSetDone) {
+        return "Nice! I've marked this task as done:\n"
+                + toSetDone.listEntry() + "\n";
+    }
```
Maybe can change javadoc that returns a string now (after changing to GUi)?

LiuZhi1an replied:
Like the "Displays" can be changed to "Returns".

LiuZhi1an commented on src/main/java/task/Deadline.java:
```diff
@@ -0,0 +1,82 @@
+     * @return true if the deadline is on the given date.
+     */
+    @Override
+    public boolean isTodayTask(LocalDate l) {
+        return l.isEqual(deadlineDate);
+    }
```
Would be more reader friendly if the parameter name is more descriptive. It might be ok for shorter methods like this, but names like dateGiven would be more intuitive.

LiuZhi1an commented:
Great job! I think there are very few violations of coding standards and your structure of classes are quite clear and well-packaged:) Only some methods are a bit long but comments were added inside the method. Perhaps helper methods could be added to write shorter main method.

ntwbruce committed:
Add some assertions to TaskList.java and Parser.java

Utilise assertions to verify the validity of important assumptions made
in the logic of the Duke program.

Finding AssertionErrors while running the program allows for fixing of
major logical errors.

Let's add a couple of assertions to the logic in the Parser and TaskList
classes.

ntwbruce committed:
Refactor the Parser class logic

Parser's utility method parse is around 150 lines long.

Abstracting out some of the logic into smaller static methods which
parse can then use will enhance code readability and make it easier to
locate bugs.

Let's refactor the parse function by:
* Abstracting out methods like parseDone, parseDelete etc
* Reducing deep nesting
* Using guard clauses to un-nest happy paths

ntwbruce committed:
Remove dead Duke logic

After implementation of the GUI, with Main, MainWindow and DialogBox
classes, Duke no longer needs its run and main functions since there
is no longer a need to support the text-based UI.

Removing the redundant code leaves less room for confusion if others
study the program.

Let's
* remove redundant methods from Duke and Ui that used to support the
text-based UI
* remove the lines in DukeTest that require Duke.main(); DukeTest needs
to be rewritten to accommodate the new GUI

ntwbruce committed:
Remove redundant Bye command related methods

Command classes have methods and fields to indicate whether or not they
are Bye commands. The GUI uses separate logic to check for Bye commands,
so the text-based UI logic becomes redundant.

With the GUI implementation, these are no longer necessary, so let's get
rid of them.

ntwbruce committed:
Fix minor errors in Parser and Storage logic

In the process of refactoring and deleting code, some more changes
need to be made to these classes that had previously been looked over.

Let's
* remove the unsupported text-based UI from Storage
* fix minor mistake in Parser

ntwbruce committed:
Utilise Java Streams in Duke

Java provides Stream functionality as a way of pipelining methods on a
sequence of objects. Methods that iterate through the task list might
be suitable for using Streams.

We thus should use the functionality provided to us by Java to improve
clarity.

Let's
* convert the logic for showList, showDateFind and showKeywordFind
methods in Ui to utilise Java Streams
* make minor changes to methods in TaskList to remove redundancies

ntwbruce committed:
Merge pull request #1 from ntwbruce/branch-A-Assertions

Add Assertions to Duke

ntwbruce committed:
Merge branch 'master' of https://github.com/ntwbruce/ip

ntwbruce committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/Parser.java

ntwbruce committed:
Merge branch 'master' into branch-A-Streams

ntwbruce committed:
Merge pull request #2 from ntwbruce/branch-A-CodeQuality

Improve code quality through refactoring and removal of redundant code

ntwbruce committed:
Merge branch 'master' of https://github.com/ntwbruce/ip

ntwbruce committed:
Merge branch 'master' into branch-A-Streams

# Conflicts:
#	src/main/java/duke/Ui.java

ntwbruce committed:
Merge pull request #3 from ntwbruce/branch-A-Streams

Add Java Streams in Duke

ntwbruce committed:
Merge branch 'master' of https://github.com/ntwbruce/ip

ntwbruce committed:
Set up Continuous Integration with GitHub Actions

ntwbruce committed:
Modify Find Date functionality

To implement a schedule viewing function, we can modify our existing
function that finds a list of tasks by date.

We want to arrange the tasks in order of their start time (for events)
or the deadline time (for deadlines). For this we can get Task classes
to implement the Comparable interface, to allow for comparing of
Deadline and Event timings.

Tasks without specified timings can be in a separate list from the timed
ones.

Let's modify Ui and the task package classes to accommodate these new
functionalities.

ntwbruce committed:
Modify MainWindow layout

Whenever the 'list' command is called in the GUI, the DialogBox display
always truncates the list if it is too long, preventing users from
seeing the whole list.

Thus let's modify the details of MainWindow to allow for the entire list
to always be displayed.

ntwbruce committed:
Convert FindDateCommand to ShowScheduleCommand

After reformatting the output of the Find Date logic to return Show
Schedule logic, convert the whole FindDateCommand class to the
ShowScheduleCommand class.

Let's
* modify all classes that utilise the former FindDateCommand class
to now use the ShowScheduleCommand class
* add a new DukeException to handle incorrect user input for finding
schedules

ntwbruce committed:
Update GUI images for Duke and User

Add transparent chicken and duck PNGs for GUI aesthetic.

ntwbruce committed:
Remove unnecessary Duke methods and DukeExceptions

After creating schedule function, MISSING_FIND_KEYWORD and
MISSING_FIND_DATE exceptions no longer used.

Duke byeMessage method not necessary as program closing message is
handled by the Ui class.

ntwbruce committed:
Improve GUI aesthetics

Add custom colours, rounded corners, and text padding to the GUI dialog
boxes. Also, add Duke custom welcome message and exit message.

Let's
* add colours to dialog boxes: blue for user messages, green for Duke
default messages and red for Duke exception messages
* add colours to the MainWindow: floral white
* add title "DUKE" to the GUI window
* customise spacing, font and font size for GUI text
* modify MainWindow to display welcome and exit messages

ntwbruce committed:
Merge branch 'branch-betterGUI'

ntwbruce committed:
Refactor Duke to prepare for renaming

Place task, exception and command packages, as well as duke classes,
inside new duck package.

ntwbruce committed:
Change every instance of Duke to Duck

ntwbruce committed:
Fix parsing bug in Storage

The load method in the Storage class is unable to correctly parse
Deadline and Event strings with descriptions longer than one word.
Make the necessary changes to allow it to parse Task strings of any
word count.

ntwbruce committed:
Change Duck GUI text

Make Duck say "QUACK!!!" instead of "OOPS!!!" when an exception is
encountered.

ntwbruce committed:
Clear hard disk

Database file should be empty by default. Likely forgot to clear it
after previous use.

ntwbruce committed:
Merge branch 'branch-RenameToDuck'

ntwbruce committed:
Set up User Guide

In preparation for release of the Duck application, write up a User
Guide to provide users a simple set of instructions for usage.

Currently the name used in the UG is Duke, but will be changed to Duck
once the updated JAR file is released on GitHub.

ntwbruce committed:
Set theme jekyll-theme-leap-day

ntwbruce committed:
Update Gradle config and User Guide

Since we changed Duke's name to Duck, modify build.gradle to now create
the JAR file duck.jar, and the User Guide to call the application Duck.

ntwbruce committed:
Fix minor typo in User Guide

ntwbruce committed:
Fix minor bugs in saving data to hard disk and Find functionality

ntwbruce committed:
Fix Storage bugs

Some BufferedReader objects were not closed in the Storage methods,
which causes File.delete() functions to not work as intended on Windows
systems.

ntwbruce committed:
Update User Guide to include links to new release

Duck has been released on the GitHub. Modify the User Guide to provide
links for users to download the Duck JAR file.

ntwbruce committed:
Minor fixes to follow Java coding standard

ntwbruce committed:
Update JUnit files

Old JUnit files DukeTest.java and StorageTest.java no longer function
as intended due to conversion of Duke from a CLI to a GUI application.

Let's remove the outdated files, and provide new, functional JUnit tests
for some of the current main Duck classes.

ntwbruce committed:
Update UG with v0.3.1 release links
